### PR TITLE
SONARJAVA-2555 debug mode and debug SE rules

### PR DESCRIPTION
**Pull request by m-g-sonar**

### CODE REVIEW

File: sonar-java-plugin/src/main/java/org/sonar/plugins/java/JavaSonarWayProfile.java [hunk 1]
```diff
@@ -30,40 +38,51 @@
...
  * Replacement for org.sonar.plugins.squid.SonarWayProfile
  */
 @SonarLintSide
 public class JavaSonarWayProfile extends ProfileDefinition {
 
   private final RuleFinder ruleFinder;
-  public JavaSonarWayProfile(RuleFinder ruleFinder) {
+  private final Settings settings;
+
+  public JavaSonarWayProfile(RuleFinder ruleFinder, Settings settings) {
```
Comment (SonarTech) [inline sonar-java-plugin/src/main/java/org/sonar/plugins/java/JavaSonarWayProfile.java]: MINOR Remove this use of "RuleFinder"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)
MINOR Remove this use of "Settings"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

File: sonar-java-plugin/src/main/java/org/sonar/plugins/java/JavaSonarWayProfile.java [hunk 2]
```diff
@@ -30,40 +38,51 @@
...
   }
 
+  private void activateRule(RulesProfile profile, String repository, String key) {
+    Rule rule = ruleFinder.findByKey(repository, key);
+    if (rule != null) {
+      profile.activateRule(rule, null);
+    }
+  }
+
+  private Stream<String> debugKeys() {
```
Comment (SonarTech) [inline sonar-java-plugin/src/main/java/org/sonar/plugins/java/JavaSonarWayProfile.java]: MAJOR Make "debugKeys" a "static" method. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2325)

File: sonar-java-plugin/src/main/java/org/sonar/plugins/java/JavaRulesDefinition.java [hunk 3]
```diff
@@ -45,6 +48,12 @@
  */
 public class JavaRulesDefinition implements RulesDefinition {
 
+  private final boolean isDebugEnabled;
+
+  public JavaRulesDefinition(Settings settings) {
```
Comment (SonarTech) [inline sonar-java-plugin/src/main/java/org/sonar/plugins/java/JavaRulesDefinition.java]: MINOR Remove this use of "Settings"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 3 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: sonar-java-plugin/src/main/java/org/sonar/plugins/java/Java.java [hunk 4]
```diff
@@ -59,6 +58,11 @@
    */
   public static final String SOURCE_VERSION = "sonar.java.source";
 
+  /**
+   * Turn ON debug rules
+   */
+  public static final String DEBUG = "sonar.java.debug";
```
Comment (benzonico) [inline sonar-java-plugin/src/main/java/org/sonar/plugins/java/Java.java]: Worth naming more precisely the property key IMO. DEBUG_RULE_KEY ?

File: sonar-java-plugin/src/main/java/org/sonar/plugins/java/JavaRulesDefinition.java [hunk 5]
```diff
@@ -90,8 +102,8 @@ protected void newRule(Class<?> ruleClass, NewRepository repository, JavaSonarWa
     }
   }
 
-  private static boolean debugRuleEnabled(Class<?> ruleClass) {
-    return DebugCheck.class.isAssignableFrom(ruleClass);
+  private boolean debugRuleEnabled(Class<?> ruleClass) {
```
Comment (benzonico) [inline sonar-java-plugin/src/main/java/org/sonar/plugins/java/JavaRulesDefinition.java]: activateDebugRule could be a more suitable name don't you think ?

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: java-frontend/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java [hunk 6]
```diff
@@ -25,7 +25,22 @@
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableSet;
 import com.google.common.collect.Lists;
-
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java]: All this is a useless modification of imports.

File: java-frontend/src/main/java/org/sonar/java/se/checks/debug/package-info.java [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.se.checks.debug;
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/se/checks/debug/package-info.java]: Missing annotation about parametersAreNonNullByDefault ?

File: java-frontend/src/test/files/se/debug/DebugMethodYieldsCheck.java [hunk 8]
```diff
@@ -14,4 +25,25 @@ Object bar(boolean b) { // Compliant - we do not have method behaviors for overr
   }
 
   abstract Object qix(boolean b); // Compliant
+
+  static void plop() { // Compliant - too many steps
```
Comment (benzonico) [inline java-frontend/src/test/files/se/debug/DebugMethodYieldsCheck.java]: This is going to be a "long" test to run and will add unecessary time for build. We should probably work out some solution to  this problem by not actually reaching the limit we use in production.

Comment (m-g-sonar) [reply java-frontend/src/test/files/se/debug/DebugMethodYieldsCheck.java]: Added a test with a parametrized max step.

**Pull request merged**